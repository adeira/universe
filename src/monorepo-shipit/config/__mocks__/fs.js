// @flow

const fs: any = jest.createMockFromModule('fs');

fs.existsSync = (path) => path !== '/unknown_path';

export default fs;
