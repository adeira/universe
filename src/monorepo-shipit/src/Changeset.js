// @flow strict

import { sprintf } from '@adeira/js';

type Diff = {
  +path: string,
  +body: string,
};

opaque type ChangesetData = {
  +id: string,
  +timestamp: string,
  +author: string,
  +subject: string,
  +description: string,
  +diffs: Set<Diff>,
  +coAuthorLines: $ReadOnlyArray<string>,
  +debugMessages: $ReadOnlyArray<string>,
};

export default class Changeset {
  declare id: string;
  declare timestamp: string;
  declare author: string;
  declare subject: string;
  declare description: string;
  declare diffs: Set<Diff>;
  coAuthorLines: $ReadOnlyArray<string> = [];
  debugMessages: $ReadOnlyArray<string> = [];

  isValid(): boolean {
    return this.diffs.size > 0;
  }

  getID(): string {
    return this.id;
  }

  withID(id: string): Changeset {
    return this.__clone({ id });
  }

  getTimestamp(): string {
    return this.timestamp;
  }

  withTimestamp(timestamp: string): Changeset {
    return this.__clone({ timestamp });
  }

  getAuthor(): string {
    return this.author;
  }

  withAuthor(author: string): Changeset {
    return this.__clone({ author });
  }

  getCoAuthorLines(): $ReadOnlyArray<string> {
    return this.coAuthorLines;
  }

  withCoAuthorLines(coAuthorLines: $ReadOnlyArray<string>): Changeset {
    return this.__clone({ coAuthorLines });
  }

  getSubject(): string {
    return this.subject;
  }

  withSubject(subject: string): Changeset {
    return this.__clone({ subject });
  }

  getDescription(): string {
    return this.description;
  }

  withDescription(description: string): Changeset {
    return this.__clone({ description });
  }

  getCommitMessage(): string {
    return `${this.getSubject()}\n\n${this.getDescription()}`;
  }

  getDiffs(): Set<Diff> {
    return this.diffs ?? new Set();
  }

  withDiffs(diffs: Set<Diff>): Changeset {
    return this.__clone({ diffs });
  }

  withDebugMessage(string: string, ...args: $ReadOnlyArray<string>): Changeset {
    const messages = this.debugMessages;
    return this.__clone({
      debugMessages: messages.concat(sprintf(string, ...args)),
    });
  }

  dumpDebugMessages(): void {
    console.log(sprintf('  DEBUG %s (%s)', this.getID(), this.getSubject()));
    for (const debugMessage of this.debugMessages) {
      console.log('    %s', debugMessage);
    }
  }

  __clone(newProps: { [$Keys<ChangesetData>]: $Values<ChangesetData>, ... }): Changeset {
    return Object.assign(Object.create(Object.getPrototypeOf(this)), this, newProps);
  }
}
