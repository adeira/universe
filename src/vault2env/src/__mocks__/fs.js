// @flow

const fs: any = jest.createMockFromModule('fs');

let mockFiles = Object.create(null);
fs.__setMockFiles = function (newMockFiles) {
  mockFiles = newMockFiles;
};

fs.existsSync = function (filePath) {
  return mockFiles[filePath] !== undefined;
};

fs.writeFileSync = function (filePath, content) {
  mockFiles[filePath] = content;
};

fs.readFileSync = function (filePath) {
  return mockFiles[filePath];
};

module.exports = fs;
