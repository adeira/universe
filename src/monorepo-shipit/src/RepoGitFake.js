// @flow

import fs from 'fs';
import os from 'os';
import path from 'path';
import { ShellCommand } from '@adeira/monorepo-utils';

import accounts from './accounts';
import RepoGit from './RepoGit';

export default class RepoGitFake extends RepoGit {
  #testRepoPath: string;

  constructor(
    testRepoPath: string = fs.mkdtempSync(path.join(os.tmpdir(), 'adeira-shipit-tests-')),
  ) {
    new ShellCommand(testRepoPath, 'git', 'init').runSynchronously();
    const username = 'adeira-shipit-tests';
    for (const [key, value] of Object.entries({
      'user.email': accounts.get(username),
      'user.name': username,
    })) {
      new ShellCommand(
        testRepoPath,
        'git',
        'config',
        key,
        // $FlowIssue[incompatible-call]: https://github.com/facebook/flow/issues/2174
        value,
      ).runSynchronously();
    }
    super(testRepoPath);
    this.#testRepoPath = testRepoPath;
  }

  push: (destinationBranch: string) => void = () => {};

  configure: () => void = () => {};

  checkoutBranch: (branchName: string) => void = () => {};

  clean: () => void = () => {};

  /* $FlowFixMe[incompatible-type] This comment suppresses an error when
   * upgrading Flow. To see the error delete this comment and run Flow. */
  /* $FlowFixMe[signature-verification-failure] This comment suppresses an
   * error when upgrading Flow. To see the error delete this comment and run
   * Flow. */
  export = (): void => {};

  getFakeRepoPath(): string {
    return this.#testRepoPath;
  }

  printFakeRepoHistory(): string {
    return this._gitCommand('log', '--stat', '--pretty=format:SUBJ: %s%nDESC: %b')
      .runSynchronously()
      .getStdout()
      .trim();
  }
}
