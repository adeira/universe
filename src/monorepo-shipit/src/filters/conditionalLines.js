// @flow strict

import Changeset from '../Changeset';

// https://github.com/lodash/lodash/blob/f8c7064d450cc068144c4dad1d63535cba25ae6d/escapeRegExp.js
function _e(s) {
  // RegExp.escape substitute
  return s.replace(/[\\^$.*+?()[\]{}|]/g, '\\$&');
}

function process(changeset: Changeset, pattern: RegExp, replacement: string) {
  const diffs = new Set();
  for (const diff of changeset.getDiffs()) {
    const newDiff = {
      ...diff,
      body: diff.body
        .split('\n')
        .map(line => line.replace(pattern, replacement))
        .join('\n'),
    };
    diffs.add(newDiff);
  }
  return changeset.withDiffs(diffs);
}

export function commentLines(
  changeset: Changeset,
  marker: string = '@x-oss-disable',
  commentStart: string = '//',
  commentEnd: null | string = null,
) {
  const ending = commentEnd === null ? '' : ` ${commentEnd}`;
  const pattern = new RegExp(`^([-+ ]\\s*)(\\S.*) ${_e(commentStart)} ${_e(marker)}${_e(ending)}$`);
  return process(changeset, pattern, `$1${commentStart} ${marker}: $2${ending}`);
}

export function uncommentLines(
  changeset: Changeset,
  marker: string = '@x-oss-disable',
  commentStart: string = '//',
  commentEnd: null | string = null,
) {
  const ending = commentEnd === null ? '' : ` ${commentEnd}`;
  const pattern = new RegExp(`^([-+ ]\\s*)${_e(commentStart)} ${_e(marker)}: (.+)${_e(ending)}$`);
  return process(changeset, pattern, `$1$2 ${commentStart} ${marker}${ending}`);
}
