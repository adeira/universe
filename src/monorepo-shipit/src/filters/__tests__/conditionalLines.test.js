// @flow

import fs from 'fs';
import path from 'path';

import { commentLines, uncommentLines } from '../conditionalLines';
import RepoGitFake from '../../RepoGitFake';
import Changeset from '../../Changeset';

function printBodies(changeset: Changeset): string {
  return [...changeset.getDiffs()].map(diff => diff.body).toString();
}

function getChangeset(patchName: string): Changeset {
  const repo = new RepoGitFake();
  const patch = fs.readFileSync(path.join(__dirname, 'fixtures', patchName), 'utf8');
  return repo.getChangesetFromExportedPatch('MOCKED', patch);
}

it('comments lines correctly - no comment end', () => {
  const changeset = getChangeset('comment-lines-no-comment-end.patch');
  const newChangeset = commentLines(changeset, '@x-oss-disable', '//', null);
  const revertedChangeset = uncommentLines(newChangeset, '@x-oss-disable', '//', null);

  expect({
    input: printBodies(changeset),
    output: printBodies(newChangeset),
  }).toMatchSnapshot();

  expect(printBodies(revertedChangeset)).toBe(printBodies(changeset));
});

it('comments lines correctly - with comment end', () => {
  const changeset = getChangeset('comment-lines-comment-end.patch');
  const newChangeset = commentLines(changeset, '@x-oss-disable', '/*', '*/');
  const revertedChangeset = uncommentLines(newChangeset, '@x-oss-disable', '/*', '*/');

  expect({
    input: printBodies(changeset),
    output: printBodies(newChangeset),
  }).toMatchSnapshot();

  expect(printBodies(revertedChangeset)).toBe(printBodies(changeset));
});

it('works with comments in comments', () => {
  const changeset = getChangeset('double-comment.patch');
  const newChangeset = commentLines(changeset);
  const revertedChangeset = uncommentLines(newChangeset);

  expect({
    input: printBodies(changeset),
    output: printBodies(newChangeset),
  }).toMatchSnapshot();

  expect(printBodies(revertedChangeset)).toBe(printBodies(changeset));
});

it('enables and disables lines as needed', () => {
  // This is a real-life example where we disable some lines but enable others.

  const changeset = getChangeset('enable-disable.patch');
  const newChangeset = uncommentLines(
    commentLines(changeset, '@x-oss-disable', '//', null),
    '@x-oss-enable',
    '//',
    null,
  );

  expect({
    input: printBodies(changeset),
    output: printBodies(newChangeset),
  }).toMatchSnapshot();
});
