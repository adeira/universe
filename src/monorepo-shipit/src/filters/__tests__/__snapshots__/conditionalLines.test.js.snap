// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comments lines correctly - no comment end 1`] = `
{
  "input": "index 1fa2ce4..499116a 100644
--- a/foo
+++ b/foo
@@ -1,7 +1,7 @@
 foo // @x-oss-disable
 normal_1
 bar // @x-oss-disable
+baz // @x-oss-disable

   herp // @x-oss-disable
   normal_2
-  derp // @x-oss-disable
",
  "output": "index 1fa2ce4..499116a 100644
--- a/foo
+++ b/foo
@@ -1,7 +1,7 @@
 // @x-oss-disable: foo
 normal_1
 // @x-oss-disable: bar
+// @x-oss-disable: baz

   // @x-oss-disable: herp
   normal_2
-  // @x-oss-disable: derp
",
}
`;

exports[`comments lines correctly - with comment end 1`] = `
{
  "input": "index 5bcc9b1..17467d2 100644
--- a/foo
+++ b/foo
@@ -1,7 +1,7 @@
 foo /* @x-oss-disable */
 normal_1
 bar /* @x-oss-disable */
+baz /* @x-oss-disable */

   herp /* @x-oss-disable */
   normal_2
-  derp /* @x-oss-disable */
",
  "output": "index 5bcc9b1..17467d2 100644
--- a/foo
+++ b/foo
@@ -1,7 +1,7 @@
 /* @x-oss-disable: foo */
 normal_1
 /* @x-oss-disable: bar */
+/* @x-oss-disable: baz */

   /* @x-oss-disable: herp */
   normal_2
-  /* @x-oss-disable: derp */
",
}
`;

exports[`enables and disables lines as needed 1`] = `
{
  "input": "index 1fa2ce4..499116a 100644
--- a/foo
+++ b/foo
@@ -1,5 +1,5 @@
 // @x-oss-enable: foo
 bar // @x-oss-disable
+baz // @x-oss-disable

   herp // @x-oss-disable
-  // @x-oss-enable: derp
",
  "output": "index 1fa2ce4..499116a 100644
--- a/foo
+++ b/foo
@@ -1,5 +1,5 @@
 foo // @x-oss-enable
 // @x-oss-disable: bar
+// @x-oss-disable: baz

   // @x-oss-disable: herp
-  derp // @x-oss-enable
",
}
`;

exports[`works with comments in comments 1`] = `
{
  "input": "index 1fa2ce4..499116a 100644
--- a/foo
+++ b/foo
@@ -1,5 +1,5 @@
 foo // first-comment // @x-oss-disable
 bar // first-comment // @x-oss-disable
+baz // first-comment // @x-oss-disable

   herp // first-comment // @x-oss-disable
-  derp // first-comment // @x-oss-disable
",
  "output": "index 1fa2ce4..499116a 100644
--- a/foo
+++ b/foo
@@ -1,5 +1,5 @@
 // @x-oss-disable: foo // first-comment
 // @x-oss-disable: bar // first-comment
+// @x-oss-disable: baz // first-comment

   // @x-oss-disable: herp // first-comment
-  // @x-oss-disable: derp // first-comment
",
}
`;
